### ISSUE TITLE
Exception safety and null pointer dereferencing

### ISSUE THREAD

**Issue by godbyk** [vrpn/vrpn#215]
This will cause a segfault if the constructor fails and silently swallows the exception (leaving `d_5dt_button` set to `NULL`):

https://github.com/vrpn/vrpn/blob/c05f12b17d3a5e93ca007074587beace606610ec/vrpn_5DT16.C#L349

Similarly for

https://github.com/vrpn/vrpn/blob/c05f12b17d3a5e93ca007074587beace606610ec/vrpn_Analog_USDigital_A2.C#L187

when `_devAddr` is set to `NULL` in the constructor upon out-of-memory.

I haven't examined all the other files yet, but I wouldn't be surprised to find a number of similar issues there.

A better approach might be to allow the device constructors to throw exceptions. Then, wrap [the huge if...else if... block](https://github.com/vrpn/vrpn/blob/master/server_src/vrpn_Generic_server_object.C#L5263) within the `vrpn_Generic_Server_Object()` constructor, in a try...catch.  This way, if a driver fails to be constructed (i.e., throws an exception), the `vrpn_Generic_Server_Object()` constructor can catch and log the exception, then continue trying to instantiate the next driver.  All the try..catch blocks in the driver constructors can be removed and we'll just let the `vrpn_Generic_Server_Object()` constructor handle them.

Note that I haven't walked through all the code yet, and it's been a while since I've read through the server code, so I may be making an embarrassingly obvious error in logic.

**Comment by russell-taylor** [comment]
Yeah, turning exceptions from construction into seg faults later is not an improvement.  Given that the constructor itself can throw an out-of-memory exception, leaving them in the constructors will be fine.  I already added the check in the generic-object class so that should catch them before the user code does.

**Comment by godbyk** [comment]
Ah, I missed the updated `VRPN_CHECK` definition. That catches exceptions thrown by the device constructors.

Would a good rule of thumb for exceptions in drivers to be:
 * Device constructors may throw exceptions if the device can't initialize properly and shouldn't be added to the list of active devices.
 * No other driver methods may throw exceptions.

In general, for libraries, I use the following rules:
 * Exceptions may be used internally in the library as per best practices.
 * Exceptions must be caught and converted to error codes at the public C API boundary.

I started fixing a few of these try...catch blocks in [my `exception-safety` branch](https://github.com/godbyk/vrpn/tree/exception-safety). Let me know if you like that direction or want to take a different tack.
